OpenSim/Simulation/Model/Model.cpp

```diff
@@ -1875,6 +1851,50 @@ bool Model::getAllControllersEnabled() const{
 void Model::setAllControllersEnabled( bool enabled ) {
     _allControllersEnabled = enabled;
 }
+bool Model::isControlsCacheValid(const SimTK::State& state) const{
+    return getDefaultSubsystem().isCacheValueRealized(state,_modelControlsIndex);
+}
+//------------------------------------------------------------------------------
+//            OVERRIDDEN METHOD TO INITIALIZE CONTROLS CACHE
+//------------------------------------------------------------------------------
+void Model::extendRealizeTopology(SimTK::State& s) const
+{    
+    Super::extendRealizeTopology(s);
+
+    Model *mutableThis = const_cast<Model *>(this);
+
+    // Create the shared cache that will hold all model controls
+    // This must be created before Actuator.extendAddToSystem() since Actuator will append 
+    // its "slots" and retain its index by accessing this cached Vector
+    // value depends on velocity and invalidates dynamics BUT should not trigger
+    // recomputation of the controls which are necessary for dynamics
+    mutableThis->_modelControlsIndex = getDefaultSubsystem().allocateCacheEntry(
+        s, Stage::Velocity, Stage::Acceleration, new Value<Vector>());
+}
+//------------------------------------------------------------------------------
+//       OVERRIDDEN METHOD TO COMPUTE CONTROLS DURING REALIZE VELOCITY
+//------------------------------------------------------------------------------
+void Model::extendRealizeVelocity(const SimTK::State& state) const
+{
+    Super::extendRealizeVelocity(state);
+
+    //Calculate the controls cache before we realize dynamics and call calcForces (possibly in parallel).
+    //Note: If the shared controls cache is calculated inside of calcForces and the force in which it is
+    //being calculated is parallel, a data race may occur to mark the controlsCache as valid/invalid.
+    if(!isControlsCacheValid(state)){
+        // Always reset controls to their default values before computing controls
```

Comment:
It shouldn't be valid in velocity stage and realizeVelocity should be called only if the stage is NOT already marked as Velocity. That means anytime this method is called the cache should be invalid

This is the opposite of a lazy evaluation. Here you are being proactive in anticipation of other stages requiring controls.

I think this if fine, but I don't think the lazy type check for validity is necessary when you are forcing the calculation for controls.